File: shared/src/main/java/ru/privatenull/pnauth/config/HashSettings.java
package ru.privatenull.pnauth.config;

import ru.privatenull.pnauth.security.HashAlgorithm;

public record HashSettings(
        HashAlgorithm algorithm,
        int pbkdf2Iterations,
        int bcryptCost,
        int argon2Iterations,
        int argon2Memory,
        int argon2Parallelism
) {
    public HashSettings {
        if (algorithm == null) {
            throw new IllegalArgumentException("algorithm must not be null");
        }
        if (pbkdf2Iterations < 10_000) {
            throw new IllegalArgumentException("pbkdf2Iterations is too small");
        }
        if (bcryptCost < 4 || bcryptCost > 31) {
            throw new IllegalArgumentException("bcryptCost must be between 4 and 31");
        }
        if (argon2Iterations < 1 || argon2Memory < 8 || argon2Parallelism < 1) {
            throw new IllegalArgumentException("Invalid Argon2 settings");
        }
    }

    public static HashSettings defaults() {
        return new HashSettings(HashAlgorithm.PBKDF2, 120_000, 12, 2, 65_536, 1);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/config/HashSettings.kt
package ru.privatenull.pnauth.config

import ru.privatenull.pnauth.security.HashAlgorithm

@JvmRecord
data class HashSettings(
    val algorithm: HashAlgorithm,
    val pbkdf2Iterations: Int,
    val bcryptCost: Int,
    val argon2Iterations: Int,
    val argon2Memory: Int,
    val argon2Parallelism: Int
) {
    init {
        requireNotNull(algorithm) { "algorithm must not be null" }
        require(pbkdf2Iterations >= 10_000) { "pbkdf2Iterations is too small" }
        require(bcryptCost in 4..31) { "bcryptCost must be between 4 and 31" }
        require(argon2Iterations >= 1 && argon2Memory >= 8 && argon2Parallelism >= 1) {
            "Invalid Argon2 settings"
        }
    }

    companion object {
        @JvmStatic
        fun defaults(): HashSettings {
            return HashSettings(HashAlgorithm.PBKDF2, 600_000, 12, 2, 65_536, 1)
        }
    }
}
